thewalrus/quantum.py

```diff
@@ -788,3 +821,77 @@ def total_photon_num_dist_pure_state(cov, cutoff=50, hbar=2, padding_factor=2):
         rs = np.arctanh(np.linalg.svd(B, compute_uv=False))
         return gen_multi_mode_dist(rs, cutoff=cutoff, padding_factor=padding_factor)[0:cutoff]
     raise ValueError("The Gaussian state is not pure")
+
+
+
+
+def choi_expand(S, r=np.arcsinh(1.0)):
+    r"""
+    Construct the Gaussian-Choi-Jamiolkowski expansion of a symplectic matrix S.
+
+    Args:
+        S (array): symplectic matrix
+
+    Returns:
+        (array): expanded symplectic matrix of twice the size of S
+    """
+
+    (n, m) = S.shape
+    if n != m:
+        raise ValueError("The matrix S is not square")
+    if n % 2 != 0:
+        raise ValueError("The matrix S is not of even size")
+
+    nmodes = n // 2
+
+    ch = np.cosh(r) * np.identity(nmodes)
+    sh = np.sinh(r) * np.identity(nmodes)
+    zh = np.zeros([nmodes, nmodes])
+    Schoi = np.block([[ch, sh, zh, zh], [sh, ch, zh, zh], [zh, zh, ch, -sh], [zh, zh, -sh, ch]])
+
+    return expand(S, list(range(nmodes)), n) @ Schoi
+
+
+def fock_tensor(S, alpha, cutoff, r=np.arcsinh(1.0), check_symplectic=True):
+    r"""
+    Calculates the Fock representation of a Gaussian unitary parametrized by
+    the symplectic matrix S and the displacements alpha up to cutoff in Fock space.
+
+    Args:
+        S (array): symplectic matrix
+        alpha (array): complex vector of displacements
+        cutoff (int): cutoff in Fock space
+        r (float): squeezing parameter used for the Choi expansion
+        check_symplectic (boolean): checks whether the input matrix is symplectic
+    Return:
+        (array): Tensor containing the Fock representation of the Gaussian unitary
+    """
+    # Check the matrix is symplectic
+    if check_symplectic:
+        if not is_symplectic(S):
+            raise ValueError("The matrix S is not symplectic")
+
+    # And that S and alpha have compatible dimensions
+    l, _ = S.shape
+    if l // 2 != len(alpha):
+        raise ValueError("The matrix S and the vector alpha do not have compatible dimensions")
+
+    # Construct its Choi expansion and state and the call state_vector
+    S_exp = choi_expand(S, r)
+    cov = S_exp @ S_exp.T
+    l = len(alpha)
+    alphat = np.array(list(alpha) + ([0] * l))
+    x = 2 * alphat.real
+    p = 2 * alphat.imag
+    mu = np.concatenate([x, p])
+    tensor = state_vector(mu, cov, normalize=False, cutoff=cutoff, hbar=2, check_purity=False)

```

Comment:
This is because `state_vector` uses `batch_hafnian` under the hood right?

Reply:
correct!